src/number_theory/arithmetic_function.lean

```diff
@@ -153,23 +159,35 @@ instance [add_comm_group R] : add_comm_group (arithmetic_function R) :=
 { .. arithmetic_function.add_comm_monoid,
   .. arithmetic_function.add_group }
 
-section dirichlet_ring
-variable [semiring R]
+section has_scalar
+variables {M : Type*} [has_zero R] [add_comm_monoid M] [has_scalar R M]
 
 /-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
   such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
-instance : has_mul (arithmetic_function R) :=
-⟨λ f g, ⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst * g x.snd, by simp⟩⟩
+instance : has_scalar (arithmetic_function R) (arithmetic_function M) :=
+⟨λ f g, ⟨λ n, ∑ x in divisors_antidiagonal n, f x.fst • g x.snd, by simp⟩⟩
 
 @[simp]
-lemma mul_apply {f g : arithmetic_function R} {n : ℕ} :
+lemma smul_apply {f : arithmetic_function R} {g : arithmetic_function M} {n : ℕ} :
+  (f • g) n = ∑ x in divisors_antidiagonal n, f x.fst • g x.snd := rfl
+
+end has_scalar
+
+/-- The Dirichlet convolution of two arithmetic functions `f` and `g` is another arithmetic function
+  such that `(f * g) n` is the sum of `f x * g y` over all `(x,y)` such that `x * y = n`. -/
+instance [semiring R] : has_mul (arithmetic_function R) := ⟨(•)⟩
+
+@[simp]
+lemma mul_apply [semiring R] {f g : arithmetic_function R} {n : ℕ} :
   (f * g) n = ∑ x in divisors_antidiagonal n, f x.fst * g x.snd := rfl
 
-instance : monoid (arithmetic_function R) :=
-{ one_mul := λ f,
-  begin
+section module
+variables {M : Type*} [semiring R] [add_comm_monoid M] [semimodule R M]
+
+lemma one_smul' (b : arithmetic_function M) : (1 : arithmetic_function R) • b = b :=
+begin
```

Comment:
I suspect this lemma and the one below it only needs `mul_action R M`

Reply:
Only one of them can be generalized in that way.

Reply:
Oops, neither can

Reply:
Yep, you're right - thanks for trying!